pip/main.ts

```diff
@@ -0,0 +1,158 @@
+//
+// This prevents us from skipping the analysis when an argument is passed before
+// the first subcommand (i.e.: `pip --no-color install package`).
+const subcommand = Deno.args[0];
+if (Deno.args.length != 0 && !knownSubcommands.includes(subcommand)) {
+  console.error(`[${red("phylum")}] This extension does not support arguments before the first subcommand. Please open an issue if "${subcommand}" is not an argument.`);
+}
+
+// Ignore all commands that shouldn't be intercepted.
+if (Deno.args.length == 0 || subcommand != "install") {
+  const cmd = Deno.run({ cmd: ["pip", ...Deno.args] });
+  const status = await cmd.status();
+  Deno.exit(status.code);
+}
+
+// Analyze new dependencies with phylum before install/update.
+await checkDryRun();
+
+// Perform the package installation.
+const cmd = Deno.run({ cmd: ["pip", ...Deno.args] });
+const status = await cmd.status();
+Deno.exit(status.code);
+
+// Analyze new packages.
+async function checkDryRun() {
+  console.log(`[${green("phylum")}] Finding new dependencies…`);
+
+  const status = PhylumApi.runSandboxed({
+    cmd: "pip",
+    args: [...Deno.args, "--dry-run"],
+    exceptions: {
+      run: ["/bin", "/usr/bin", "~/.pyenv"],
+      write: ["~/Library/Caches/pip", "~/.cache", "~/.local/lib", "/tmp"],
+      read: ["~/Library/Caches/pip", "~/.cache", "~/.local/lib", "/tmp", "/etc/passwd", "/etc/apache2/mime.types"],
+      net: true,
+    },
+    stdout: 'piped',
+  });
+
+  // Ensure dry-run was successful.
+  if (!status.success) {
+    console.error(`[${red("phylum")}] Pip dry-run failed.\n`);
+    await Deno.exit(status.code);
+  }
+
+  // Parse dry-run output.
+  let packages = parseDryRun(status.stdout);
+
+  console.log(`[${green("phylum")}] Dependency resolution successful.\n`);
+  console.log(`[${green("phylum")}] Analyzing packages…`);
+
+  if (packages.length === 0) {
+    console.log(`[${green("phylum")}] No new packages found for analysis.\n`);
+    return;
+  }
+
+  const jobId = await PhylumApi.analyze("pypi", packages);
+  const jobStatus = await PhylumApi.getJobStatus(jobId);
+
+  if (jobStatus.pass && jobStatus.status === "complete") {
+    console.log(`[${green("phylum")}] All packages pass project thresholds.\n`);
+  } else if (jobStatus.pass) {
+    console.warn(
+      `[${
+        yellow(
+          "phylum",
+        )
+      }] Unknown packages were submitted for analysis, please check again later.\n`,
+    );
+    Deno.exit(126);
+  } else {
+    console.error(
+      `[${red("phylum")}] The operation caused a threshold failure.\n`,
+    );
+    Deno.exit(127);
+  }
+}
+
+// Parse the dry-run output of `pip install`.
+function parseDryRun(output: string): [object] {
+  // Extract the "Would install [..]" line.
+  let new_deps;
+  const lines = output.split('\n');
+  for (var i = 0; i < lines.length; i++) {
+    if (lines[i].startsWith('Would install ')) {
+      new_deps = lines[i].substring('Would install '.length);
+      break;
```

Comment:
We're in some cases executing untrusted code in the sandbox to determine the package dependencies, right? Is it possible an attacker adds `print('Would install pytest 7.2.0')` to `setup.py` and then this code would determine that _only_ pytest_ 7.2.0 is being installed and allow the attacker's package to be installed?

Is it possible the attacker could output their own fake "Would install" and then also suppress the real "Would install" (redirect the process stdout?) such that we have no way to know what would actually be installed? Should we be doing the dry run and then installing specifically the packages we think we are installing instead of passing through the original arguments?

Reply:
We shouldn't be installing anything. This is running `--dry-run`.

Reply:
We're doing dry-run now, but later we are installing whatever would be installed by the same arguments.

Reply:
At least on my machine, pip runs the setup.py in a subprocess to generate an egg info file in a temp directory. So an attacker could detect that they're in the sandbox and pick different dependencies, but they couldn't write their own "Would install" line to hide the immediate package from the output.

Reply:
Are you saying it's running `setup.py` during `--dry-run` or in the install later?

It's correct that we're running the install later, but at that point the output doesn't matter anymore.

I'm not aware of what is and isn't possible in a `setup.py` unfortunately.

Reply:
`setup.py` is run during both the dry run and the installation. I think we need to install exactly the packages we think we're installing when we do the installation. Otherwise somebody can do something like this:

```py
#!/usr/bin/env python

from distutils.core import setup

import os
import os.path

try:
    os.path.lexists(os.path.expanduser("~/.ssh/id_rsa"))
except:
    # We don't have permission to read ~ or ~/.ssh, so assume sandboxed.
    setup(name="nice", version="1.0", packages=[])
    os.exit(0)

setup(
    name="evil",
    version="1.0",
    packages=[],
)

```

Reply:
The `--dry-run` option was added to the `pip install` command [in pip v22.2, which was only last Summer](https://pip.pypa.io/en/stable/news/#v22-2). The option text for states:

> Don’t actually install anything, just print what would be. Can be used in combination with --ignore-installed to ‘resolve’ the requirements.

Unfortunately, that does not mean code _won't or can't_ run. It will, for source distributions (sdists aka `.tar.gz` archives). Here are some references about this topic/issue:

* [Research Post: "Automatic Execution of Code Upon Package Download on Python Package Manager"](https://medium.com/checkmarx-security/automatic-execution-of-code-upon-package-download-on-python-package-manager-cd6ed9e366a8)
* [Blog Post from a security researcher](https://moyix.blogspot.com/2022/09/someones-been-messing-with-my-subnormals.html)
  * Search for `--dry-run` to get to the good parts
* `pip` issue #1884: [Avoid generating metadata in `pip download --no-deps ...`](https://github.com/pypa/pip/issues/1884)
  * This applies to the `--dry-run` option as well...like for `pip install`

Reply:
> Otherwise somebody can do something like this

Since we run the sandbox for both the dry-run and the wet-run, this shouldn't affect us though, right? In both scenarios our user will be in a sandbox and thus only take the non-malicious install path. So we've successfully protected the user?

Also I'm not sure it really matters which packages are installed by a malicious package, since the package itself is already fully capable of doing whatever the malicious package might have done. So if we aren't detecting the package itself, we'd already be pwned.

Reply:
The only way it could be useful is if their real payload is detected by us and their package that changes dependencies when sandboxed is not, and even then the real payload would need to be dependent on a manual post installation action such as importing the package to get it running outside the sandbox.